docs/connectors/postgresql.asciidoc

```diff
@@ -2,5 +2,1044 @@
+2. Obtains a `SHARE UPDATE EXCLUSIVE MODE` lock on each of the monitored tables to ensure that no structural changes can occur to any of the tables while the snapshot is taking place. Note that these locks do not prevent table `INSERTS`, `UPDATES` and `DELETES` from taking place during the operation
+3. Reads the current position in the server's transaction log.
+4. Scans all of the database tables and schemas and generates on the appropriate table-specific Kafka topics `READ` events for each row.
+5. Commits the transaction.
+6. Record in the connector offsets that the connector successfully completed the snapshot.
+
+If the connector fails, is rebalanced, or stops before the snapshot is complete, the connector will begin a new snapshot when it is restarted. Once the connector does complete its initial snapshot, the PostgreSQL connector then continues streaming from the position read during step 3, ensuring that it does not miss any updates. If the connector stops again for any reason, upon restart it will simply continue streaming changes from where it previously left off. However, if the connector remains stopped for long enough, PostgreSQL might purge older WAL segments and the connector's last known position may be lost. In this case, when the connector configured with *initial* snapshot mode (the default) is finally restarted, the PostgreSQL server will no longer have the starting point and the connector will not be able to relay the changes that are not available in the write ahead log.
+
+A second snapshot mode allows the connector to perform snapshots *always*. This behaviour tells the connector to always perform a snapshot when it starts up and only after the snapshot is completed to continue streaming changes from step 3 in the above sequence. This mode can be used in cases when it's known that some WAL segments have been deleted and are no longer available, or in case of a cluster failure after a new master has been promoted so that the connector doesn't miss out on any potential changes that could've taken place after the new master had been promoted but before the connector was restarted on the new master.
+
+The third snapshot mode ensures the connector *never* performs snapshots. When a new connector is configured this way, if will either continue streaming changes from a previous stored offset or it will start from the point in time when the PostgreSQL logical replication slot was first created on the server.
+
+The final snapshot mode - *initial only* - will simply perform a database snapshot and stop streaming any other changes afterwards. If a previous snapshot was in effect but had not completed successfully, the connector will restart the snapshot process each time until this completes successfully.
+
```

Comment:
there is no easy way to tell is an `LSN` number exists or not on the server.

The way logical replication streaming works is you give the server an `LSN` and it will simply either use it (if it still has the segments) or will silently discard it and start streaming changes from the latest available `LSN`. There is no "feedback" in the logical replication streaming protocol to inform the client that the `LSN` is no longer available.

This could be implemented (probably requires looking in some system tables), but it also means changes to the connector and some additional database queries and logic. So if the `when_needed` mode is required, I'd rather have that as a feature request.